wandb/sdk/launch/agent/agent.py

```diff
@@ -0,0 +1,176 @@
+"""
+Implementation of launch agent.
+"""
+
+import os
+import sys
+import time
+from typing import Any, Dict, Iterable, List, Optional
+
+import wandb
+from wandb import Settings
+from wandb.apis.internal import Api
+from wandb.errors import LaunchException
+
+from .._project_spec import create_project_from_spec, fetch_and_validate_project
+from ..runner.abstract import AbstractRun, State
+from ..runner.loader import load_backend
+from ..utils import (
+    _is_wandb_local_uri,
+    PROJECT_DOCKER_ARGS,
+    PROJECT_SYNCHRONOUS,
+)
+
+
+def _convert_access(access: str) -> str:
+    """Converts access string to a value accepted by wandb"""
+    access = access.upper()
+    assert (
+        access == "PROJECT" or access == "USER"
+    ), "Queue access must be either project or user"
+    return access
+
+
+class LaunchAgent(object):
+    """
+    Launch agent class which polls run given run queues and launches runs for wandb launch.
+    """
+
+    STATE_MAP: Dict[str, State] = {}
+
+    def __init__(self, entity: str, project: str, queues: Iterable[str] = None):
+        self._entity = entity
+        self._project = project
+        self._api = Api()
+        self._settings = Settings()
+        self._base_url = self._api.settings().get("base_url")
+        self._jobs: Dict[int, AbstractRun] = {}
+        self._ticks = 0
+        self._running = 0
+        self._cwd = os.getcwd()
+        self._namespace = wandb.util.generate_id()
+        self._access = _convert_access("project")
+        self._queues: List[str] = []
+        self.setup_run_queues(queues)
+
+    def setup_run_queues(self, queues: Optional[Iterable[str]]) -> None:
+        """
+        Checks the project to ensure run queues exist then adds them to a list to be watched by the agent
+        """
+        project_run_queues = self._api.get_project_run_queues(
```

Comment:
Let's at least add a comment to make the queue filtering happen on the server instead in the client as it will eventually become a problem.

Reply:
Agreed. Filing a ticket and adding a comment